### PR TITLE
External analysis - fix 504

### PR DESCRIPTION
**Pull request by hanars**
For 20,000+ row file, speeds up response time from 80 seconds to 10 seconds

### CODE REVIEW

File: seqr/views/apis/summary_data_api.py [hunk 1]
```diff
@@ -113,29 +113,34 @@ def bulk_update_family_analysed_by(request):
...
+    families = set()
+    requested_families = set()
+    for row in families_data:
+        projects.add(row['project'])
+        families.add(row['family'])
+        requested_families.add((row['project'], row['family']))
+
+    family_db_id_lookup = {
+        (f['project__name'], f['family_id']): f['id'] for f in Family.objects.filter(
+            family_id__in=families, project__name__in=projects).values('id', 'family_id', 'project__name')
```
Comment (ShifaSZ) [inline seqr/views/apis/summary_data_api.py:126]: How about using `Concat` to match the `project name` and `family id` pairs?

Comment (hanars) [reply seqr/views/apis/summary_data_api.py:126]: I'd rather explicitly use tuples though, what is the reason why formatted string would be better?

File: seqr/views/apis/summary_data_api.py [hunk 2]
```diff
@@ -113,29 +113,34 @@ def bulk_update_family_analysed_by(request):
...
+    missing_from_db = requested_families - set(family_db_id_lookup.keys())
+    if missing_from_db:
+        missing_families = ', '.join([f'{family} ({project})' for project, family in sorted(missing_from_db)])
         warnings.append(f'No match found for the following families: {missing_families}')
 
     analysed_by_models = [
-        FamilyAnalysedBy(family=family, data_type=data_type, last_modified_date=datetime.now())
-        for family in families
+        FamilyAnalysedBy(family_id=family_db_id_lookup[family_key], data_type=data_type, last_modified_date=datetime.now())
+        for family_key in requested_families if family_key in family_db_id_lookup
```
[the remark this answers is not in the document]
Comment (hanars) [reply seqr/views/apis/summary_data_api.py:137]: it does though? `family_key` is the project-family_id tuple

File: seqr/views/apis/summary_data_api.py (hunk 1 above)
Comment (ShifaSZ) [reply seqr/views/apis/summary_data_api.py:126]: I mean using the Django `Concat` to connect two fields and filter out extra families. It makes the following code cleaner.

Comment (hanars) [reply seqr/views/apis/summary_data_api.py:126]: ohh you mean in the filter input not the values output. Yeah I like that!

File: seqr/views/apis/summary_data_api.py [hunk 3]
```diff
@@ -113,29 +114,36 @@ def bulk_update_family_analysed_by(request):
...
-    for f_q in family_qs[1:]:
-        family_filter_q |= f_q
-    families = Family.objects.filter(family_filter_q)
+    projects = set()
+    families = set()
+    requested_families = set()
+    for row in families_data:
+        projects.add(row['project'])
+        families.add(row['family'])
+        requested_families.add((row['project'], row['family']))
```
[the remark this answers is not in the document]
Comment (hanars) [reply seqr/views/apis/summary_data_api.py:123]: No. Its more accurate to use tuples than formatted strings, the set substraction of the tuples is a much neater way of determining whats missing, for the missing message thats not how they are formatted so we would have to split and reformat the string, and also there is no disadvantage to the tuples

File: seqr/views/apis/summary_data_api.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hanars) [reply seqr/views/apis/summary_data_api.py:140]: And of course there can be missing families, we literally have a missing handler right above this

**Review by ShifaSZ — APPROVED** [review]

**Pull request merged**